components/renku_data_services/message_queue/converters.py

```diff
@@ -280,52 +281,42 @@ class EventConverter:
+    def to_events(input: _T, event_type: type[AvroModel] | type[events.AmbiguousEvent]) -> list[Event]:
         """Generate an event for a data service model based on an event type."""
-        if isinstance(input, project_models.Project):
-            return _ProjectEventConverter.to_events(input, event_type)
-        elif isinstance(input, project_models.ProjectUpdate):
-            return _ProjectEventConverter.to_events(input.new, event_type)
-        elif isinstance(input, (user_models.UserInfo, user_models.UserWithNamespace)):
-            return _UserEventConverter.to_events(input, event_type)
-        elif input is None and event_type == type(v2.UserRemoved):  # type: ignore[unreachable]
-            # NOTE: The user that was supposed to be removed is not in the database at all, so dont send the event
-            # The code is definitely reachable it is just that mypy thinks it is not
-            return []  # type: ignore[unreachable]
-        elif (
-            isinstance(input, list)
-            and len(input) > 0
-            and all([isinstance(i, authz_models.MembershipChange) for i in input])
-        ):
-            if event_type == AmbiguousEvent.PROJECT_MEMBERSHIP_CHANGED:
-                return _ProjectAuthzEventConverter.to_events(input)
-            elif event_type == AmbiguousEvent.GROUP_MEMBERSHIP_CHANGED:
-                return _GroupAuthzEventConverter.to_events(input)
-            else:
-                raise errors.ProgrammingError(
-                    message="Found unknown event type when processing membership change list in the message queue "
-                    f"decorator, {type(event_type)}"
-                )
-        elif isinstance(input, list) and len(input) == 0:
+        if not input:
             return []
-        elif (
-            isinstance(input, user_models.UserWithNamespaceUpdate)
-            and event_type == AmbiguousEvent.UPDATE_OR_INSERT_USER
-        ):
-            return _UserEventConverter.to_events(input, event_type)
-        elif (
-            isinstance(input, list)
-            and event_type == AmbiguousEvent.INSERT_USER_NAMESPACE
-            and len(input) > 0
-            and all([isinstance(i, user_models.UserWithNamespace) for i in input])
-        ):
-            output: list[Event] = []
-            for namespace in input:
-                output.extend(_UserEventConverter.to_events(namespace, event_type))
-            return output
-        elif isinstance(input, Group):
-            return _GroupEventConverter.to_events(input, event_type)
-        else:
-            raise errors.EventError(
-                message=f"Trying to convert an unknown model of type {type(input)} to an event type {event_type}"
-            )
+
+        match event_type:
+            case v2.ProjectCreated | v2.ProjectRemoved:
+                project = cast(project_models.Project, input)
+                return _ProjectEventConverter.to_events(project, event_type)
```

Comment:
```suggestion
            case v2.ProjectCreated | v2.ProjectRemoved if isinstance(input, project_models.Project):
                return _ProjectEventConverter.to_events(input, event_type)
```
You can use a guard like that in every case so that you do not have to cast and you are sure that you have the right type.